wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/EventTreeNode.java

```diff
@@ -0,0 +1,186 @@
+ *
+ * Contributors:
+ *   Codenvy, S.A. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.che.api.vfs.impl.file.event;
+
+import com.google.common.annotations.Beta;
+
+import org.eclipse.che.api.project.shared.dto.event.FileWatcherEventType;
+import org.eclipse.che.api.vfs.impl.file.event.LoEvent.ItemType;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import static java.util.Optional.empty;
+import static java.util.stream.Stream.concat;
+import static org.eclipse.che.api.vfs.impl.file.event.LoEvent.ItemType.DIR;
+import static org.eclipse.che.api.vfs.impl.file.event.LoEvent.ItemType.FILE;
+import static org.eclipse.che.api.vfs.impl.file.event.LoEvent.ItemType.UNDEFINED;
+
+/**
+ * Virtual file system event tree is a data structure to store low level VFS
+ * events in a specific manner. Events are stored in a such way that an event
+ * tree corresponds to a file system items tree but includes only those tree
+ * branches that contain modified file system items.
+ * <p>
+ *     By design event tree represents a time-based event segment. All events
+ *     close enough (in terms of time line) to each other are included to the
+ *     tree.
+ * </p>
+ * <p>
+ *     This class uses {@link LinkedHashMap} to store an event chain, not to
+ *     lose data when we have several sequential modifications of a single
+ *     tree item.
+ * </p>
+ * <p>
+ *     Note: for convenience there is a predefined event tree root node - '/',
+ *     which is defined by {@link EventTreeNode#ROOT_NODE_NAME} constant.
+ *     All trees must be started from that node using corresponding factory
+ *     method {@link EventTreeNode#newRootInstance()}
+ * </p>
+ *
+ * @author Dmitry Kuleshov
+ *
+ * @since 4.5
+ */
+@Beta
+public class EventTreeNode {
+
+    /**
+     * Tree root node name. All properly constructed trees must be started from this node.
+     */
+    private static final String ROOT_NODE_NAME = "/";
+
+    private List<EventTreeNode>             children;
+    private String                          name;
+    private String                          path;
+    private ItemType                        type;
+    /**
+     * Event chain to store all events occurred within a single time
+     * segment with this event tree node instance.
+     * Key - timestamp in millis, value - event type
+     */
+    private Map<Long, FileWatcherEventType> events;
+
+    private EventTreeNode() {
+        this.events = new LinkedHashMap<>();
+        this.children = new LinkedList<>();
+        this.type = UNDEFINED;
+    }
+
+    public static EventTreeNode newRootInstance() {
+        return EventTreeNode.newInstance().withName(ROOT_NODE_NAME);
+    }
+
+    public static EventTreeNode newInstance() {
```

Comment:
Well dude, there is a lot of information over the internet, just type something like 'static factory method vs constructor java', so I don't think I want to spend time explaining you that, but you can get general idea by reading this [article](https://jlordiales.wordpress.com/2012/12/26/static-factory-methods-vs-traditional-constructors). And just subjectively, I like how it looks in the code.